crates/vm/levm/src/precompiles.rs

```diff
@@ -232,12 +241,106 @@ pub fn ripemd_160(
     Ok(Bytes::from(output))
 }
 
-fn modexp(
-    _calldata: &Bytes,
-    _gas_for_call: U256,
-    _consumed_gas: &mut U256,
+/// Returns the result of the modexp operation given the input parameters
+pub fn modexp(
+    calldata: &Bytes,
+    gas_for_call: U256,
+    consumed_gas: &mut U256,
 ) -> Result<Bytes, VMError> {
-    Ok(Bytes::new())
+    // If calldata does not reach the required length, we should fill the rest with zeros
+    let calldata = fill_with_zeros(calldata, 96)?;
+
+    let b_size: U256 = calldata
+        .get(0..32)
+        .ok_or(PrecompileError::ParsingInputError)?
+        .into();
+
+    let e_size: U256 = calldata
+        .get(32..64)
+        .ok_or(PrecompileError::ParsingInputError)?
+        .into();
+
+    let m_size: U256 = calldata
+        .get(64..96)
+        .ok_or(PrecompileError::ParsingInputError)?
+        .into();
+
+    if b_size == U256::zero() && m_size == U256::zero() {
+        *consumed_gas = consumed_gas
+            .checked_add(U256::from(MODEXP_STATIC_COST))
+            .ok_or(OutOfGasError::ConsumedGasOverflow)?;
+        return Ok(Bytes::new());
+    }
+
+    // Because on some cases conversions exploded before the if above
+    let b_size = usize::try_from(b_size).map_err(|_| PrecompileError::ParsingInputError)?;
+    let e_size = usize::try_from(e_size).map_err(|_| PrecompileError::ParsingInputError)?;
+    let m_size = usize::try_from(m_size).map_err(|_| PrecompileError::ParsingInputError)?;
+
+    let base_limit = b_size
+        .checked_add(96)
+        .ok_or(InternalError::ArithmeticOperationOverflow)?;
+
+    let exponent_limit = e_size
+        .checked_add(base_limit)
+        .ok_or(InternalError::ArithmeticOperationOverflow)?;
+
+    // The reason I use unwrap_or_default is to cover the case where calldata does not reach the required
+    // length, so then we should fill the rest with zeros. The same is done in modulus parsing
+    let b = calldata.get(96..base_limit).unwrap_or_default();
+    let base = BigUint::from_bytes_be(b);
+
+    let e = calldata.get(base_limit..exponent_limit).unwrap_or_default();
+    let exponent = BigUint::from_bytes_be(e);
+
+    let m = match calldata.get(exponent_limit..) {
+        Some(m) => {
+            let m_extended = fill_with_zeros(&Bytes::from(m.to_vec()), m_size)?;
+            m_extended.get(..m_size).unwrap_or_default().to_vec()
+        }
+        None => Default::default(),
+    };
+    let modulus = BigUint::from_bytes_be(&m);
+
+    let gas_cost = modexp_cost(&exponent, b_size, e_size, m_size)?;
+    increase_precompile_consumed_gas(gas_for_call, gas_cost, consumed_gas)?;
+
+    let result = mod_exp(base, exponent, modulus);
```

Comment:
If mod_exp does `(b**e) % m`, why can't we do it like the opcodes `op_exp` and `op_mod` combined?

Reply:
It's because `mod_exp` applies the modulus during the exponentiation process, since calculate exp and mod separately can first compute a massive number, making it slow and impractical for large inputs. It's like applying modulus on each step of exponentiation.